File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.js
function solution(s){
    if(s.length === 1 || s[0] === ")") return false;
    
    let cnt = 0; 
    for(let i =0; i < s.length; i++){
      s[i]==="(" ? cnt += 1 : cnt -= 1
      if(cnt < 0) break;
    }
   return cnt === 0  ? true : false
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        Stack<Character> stackStr = new Stack<>();
        
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            
            if(isOpenBracket(c)) stackStr.push(c);
            else if(stackStr.isEmpty() || !isMatchingPair(stackStr.pop(), c)) return false;
            
        }
        return stackStr.isEmpty();
    }
    
       private boolean isOpenBracket(char c) {
        return c == '(' || c == '[' || c == '{';
    }
    
        private boolean isMatchingPair(char open, char close){
        return (open == '(' && close == ')') ||
               (open == '[' && close == ']') ||
               (open == '{' && close == '}');
    }
}